README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# genomics

`genomics` is specially designed for genomic teaching. The package will
help with the generation of random nucleotides sequences with start and
stop codons (`rand_nuc_seq()`), interpretation of genetic code in
different translation tables (`gen_code()` and `aminoacid()`),
antiparallel sequences (`antiparallel()`), and RNA translation to
polypeptide chains (`translate()`, `dna_to_rna()`, `rna_to_dna()`)

# Installation

`genomics` is only on
[GitHub](https://github.com/TiagoOlivoto/genomics), and can be installed
with

``` r
devtools::install_github("TiagoOlivoto/genomics")
```

*Note*: If you are a Windows user, you should also first download and
install the latest version of
[Rtools](https://cran.r-project.org/bin/windows/Rtools/).

# Genetic code

``` r
library(genomics)
# |===========================================|
# | Tools for genomics (genomics 0.1.0)   |
# | Author: Tiago Olivoto                     |
# |===========================================|
aminoacid("AUG")
# [1] "M"
aminoacid(c("AUG", "CGA", "UGA"), type = "name")
# [1] "Methionine" "Arginine"   "*"

# standard genetic code
gen_code()
#    fb sb tb codon olc tlc          name
# 1   A  A  A   AAA   K Lys        Lysine
# 2   A  A  U   AAU   N Asn    Asparagine
# 3   A  A  G   AAG   K Lys        Lysine
# 4   A  A  C   AAC   N Asn    Asparagine
# 5   A  U  A   AUA   I Ile    Isoleucine
# 6   A  U  U   AUU   I Ile    Isoleucine
# 7   A  U  G   AUG   M Met    Methionine
# 8   A  U  C   AUC   I Ile    Isoleucine
# 9   A  G  A   AGA   R Arg      Arginine
# 10  A  G  U   AGU   S Ser        Serine
# 11  A  G  G   AGG   R Arg      Arginine
# 12  A  G  C   AGC   S Ser        Serine
# 13  A  C  A   ACA   T Thr     Threonine
# 14  A  C  U   ACU   T Thr     Threonine
# 15  A  C  G   ACG   T Thr     Threonine
# 16  A  C  C   ACC   T Thr     Threonine
# 17  U  A  A   UAA   *   *             *
# 18  U  A  U   UAU   Y Tyr     Tyrosisne
# 19  U  A  G   UAG   *   *             *
# 20  U  A  C   UAC   Y Tyr     Tyrosisne
# 21  U  U  A   UUA   L Leu       Leucine
# 22  U  U  U   UUU   F Phe Phenylalanine
# 23  U  U  G   UUG   L Leu       Leucine
# 24  U  U  C   UUC   F Phe Phenylalanine
# 25  U  G  A   UGA   *   *             *
# 26  U  G  U   UGU   C Cys      Cysteine
# 27  U  G  G   UGG   W Trp   Tryptophane
# 28  U  G  C   UGC   C Cys      Cysteine
# 29  U  C  A   UCA   S Ser        Serine
# 30  U  C  U   UCU   S Ser        Serine
# 31  U  C  G   UCG   S Ser        Serine
# 32  U  C  C   UCC   S Ser        Serine
# 33  G  A  A   GAA   E Glu  GlutamicAcid
# 34  G  A  U   GAU   D Asp  AsparticAcid
# 35  G  A  G   GAG   E Glu  GlutamicAcid
# 36  G  A  C   GAC   D Asp  AsparticAcid
# 37  G  U  A   GUA   V Val        Valine
# 38  G  U  U   GUU   V Val        Valine
# 39  G  U  G   GUG   V Val        Valine
# 40  G  U  C   GUC   V Val        Valine
# 41  G  G  A   GGA   G Gly       Glycine
# 42  G  G  U   GGU   G Gly       Glycine
# 43  G  G  G   GGG   G Gly       Glycine
# 44  G  G  C   GGC   G Gly       Glycine
# 45  G  C  A   GCA   A Ala       Alanine
# 46  G  C  U   GCU   A Ala       Alanine
# 47  G  C  G   GCG   A Ala       Alanine
# 48  G  C  C   GCC   A Ala       Alanine
# 49  C  A  A   CAA   Q Gln     Glutamine
# 50  C  A  U   CAU   H His     Histidine
# 51  C  A  G   CAG   Q Gln     Glutamine
# 52  C  A  C   CAC   H His     Histidine
# 53  C  U  A   CUA   L Leu       Leucine
# 54  C  U  U   CUU   L Leu       Leucine
# 55  C  U  G   CUG   L Leu       Leucine
# 56  C  U  C   CUC   L Leu       Leucine
# 57  C  G  A   CGA   R Arg      Arginine
# 58  C  G  U   CGU   R Arg      Arginine
# 59  C  G  G   CGG   R Arg      Arginine
# 60  C  G  C   CGC   R Arg      Arginine
# 61  C  C  A   CCA   P Pro       Proline
# 62  C  C  U   CCU   P Pro       Proline
# 63  C  C  G   CCG   P Pro       Proline
# 64  C  C  C   CCC   P Pro       Proline
```

# Generate a DNA sequence

``` r
# 'code' DNA
dna <- rand_nuc_seq(n = 40)
# Start codon  : 3-5
# Stop codon   : TAG
# Position stop: 36-38
dna
# [1] "ATATGGGCGGAGAGGGGAGCGCGAGCCGATGGCTCTAGAC"

# transcription to mRNA
mrna <- dna_to_rna(dna)
mrna
# $c_dna
# [1] "ATATGGGCGGAGAGGGGAGCGCGAGCCGATGGCTCTAGAC"
# 
# $m_dna
# [1] "TATACCCGCCTCTCCCCTCGCGCTCGGCTACCGAGATCTG"
# 
# $m_rna
# [1] "AUAUGGGCGGAGAGGGGAGCGCGAGCCGAUGGCUCUAGAC"
```

# Translation

In the following example, the mRNA (`m_rna`) in `mrna` is translated to
a polypeptide chain.

``` r
# one letter code
# stop codon as *
translate(mrna$m_rna)
# [1] "MGGEGSASRWL*"

# Three letter code
# translation continues after stop codon
translate(mrna$m_rna,
          type = "tlc",
          no_stop = TRUE,
          sep = ".")
# [1] "Met.Gly.Gly.Glu.Gly.Ser.Ala.Ser.Arg.Trp.Leu.*"
```


NEWS.md
# genomics 0.1.0
* First version of package.
* Added a `NEWS.md` file to track changes to the package.


R/get_rna.R
#' Tools for transcription DNA to RNA
#'
#' @param dna,rna A character vector containing the code sequence (e.g., "ACG")
#' @param type The type of sequence. It must be either `"code"` (the code helix)
#'   or `"mold"` the helix read by RNA polymerase.
#' @return
#' * `dna_to_rna()` returns a list with the following elements:
#'   * `c_dna` The 'code' DNA.
#'   * `m_dna` The molder DNA (antiparallel of `c_dna`)
#'   * `m_rna` The pré- messenger RNA, i.e., the `c_dna` with `U` instead `T`.
#' * `rna_to_dna()` returns a list with the following elements:
#'   * `m_rna` The messenger RNA declared.
#'   * `m_dna` The molder DNA (antiparallel of `m_rna`)
#'   * `c_rna` The 'code' DNA., i.e., the `m_rna` with `T` instead `U`.
#' @export
#' @name dna_rna
#'
#' @examples
#' # convert a dna (code or mold) to a messenger RNA
#' dna_to_rna(c("ATGCGC", "AGC"))
#' dna_to_rna(c("ATGCGC", "AGC"), type = "mold")
#'
#' # convert a messenger RNA to dna code and mold
#' rna_to_dna(c("AUGCGC", "CGACGGA"))
dna_to_rna <- function(dna, type = "code"){
  if(length(dna) > 1){
    do.call(rbind, lapply(dna, dna_to_rna, type))
  } else{
    dnac_to_rna <- function(dna_c){
      dna_c <- toupper(dna_c)
      lapply(strsplit(dna_c, "")[[1]], function(x){
        ifelse(x == "T", "U", x)
      }
      ) %>%
        as.character() %>%
        paste0(collapse = "")
    }
    if(!is_dna(dna, verbose = FALSE)){
      stop("An Uracil (U) was found in the sequence. Nothing made.", call. = FALSE)
    }
    if(type == "code"){
      dna <- dna %>% toupper()
    } else{
      dna <- antiparallel(dna) %>% toupper()
    }

    m_dna <- antiparallel(dna)
    m_rna <- dnac_to_rna(dna)
    return(list(c_dna = dna,
                m_dna = m_dna,
                m_rna = m_rna))
  }
}

#' @export
#' @name dna_rna
rna_to_dna <- function(rna){
  if(length(rna) > 1){
    do.call(rbind, lapply(rna, rna_to_dna))
  } else{
    rna <- toupper(rna)
    rna_to_cdna <- function(mrna){
      mrna <- toupper(mrna)
      lapply(strsplit(mrna, "")[[1]], function(x){
        ifelse(x == "U", "T", x)
      }
      ) %>%
        as.character() %>%
        paste0(collapse = "")
    }
    if(!is_rna(rna, verbose = FALSE)){
      stop("A Thimine (T) was found in the sequence. Nothing made.", call. = FALSE)
    }
    c_dna <- rna_to_cdna(rna)
    m_dna <- antiparallel(c_dna)
    return(list(m_rna = rna,
                c_dna = c_dna,
                m_dna = m_dna))
  }
}


R/translate.R
#' Translate a RNA sequence to a protein sequence
#'
#' Translate a nucleotide (RNA) sequence to a protein chain according to the
#' standard genetic code
#'
#' @param mrna The messenger RNA.
#' @param transl_table The translation table used. Defaults to 1 (standard
#'   genetic code). See more at [aminoacid()]
#' @param start_codon The start codon. Defaults to `"AUG"`.
#' @param start_codon_first Start the translation with the `start_codon`?
#'   defaults to `TRUE`. In this case, all bases before the `start_codon` will
#'   be ignored.
#' @param no_stop A logical indicating if the translation shouldn't stop when a
#'   stop codon is encountered. Defaults to FALSE. If `TRUE`, translation will
#'   continue up to the end of the chain, and stop codons will be represented by
#'   `"*"`.
#' @param type The type of output. One of
#' * `"olc"` for one-letter code, e.g., "V" (default).
#' * `"tlc"` for three-letter code, e.g., "Val".
#' * `"name"` for the complete name, e.g., "Valine".
#' @param sep The separator between amino acids. Defaults to `"-"`.
#' @return A character
#' @export
#'
#' @examples
#' code <- rand_nuc_seq(25, type = "rna")
#' translate(code)
translate <- function(mrna,
                      transl_table = 1,
                      start_codon = "AUG",
                      start_codon_first = TRUE,
                      no_stop = FALSE,
                      type = "olc",
                      sep = ""){
  if (length(mrna) > 1){
    as.character(do.call(rbind, lapply(mrna, translate, transl_table, start_codon, start_codon_first, no_stop, type, sep)))
  } else{
    if(!is_rna(mrna, verbose = FALSE)){
      stop("A Thimine (T) was found in the sequence. Translation interrupted.",  call. = FALSE)
    }
    mrna <- tidy_sequence(mrna)
    if (start_codon_first == TRUE){
      scod <- str_locate(mrna, start_codon)
      subs <- str_sub(mrna, min(scod), nchar(mrna))
      subs <-  split_each(subs, n = 3)
    } else{
      subs <-  split_each(mrna, n = 3)
    }
    seq <- aminoacid(subs, type = type, transl_table = transl_table)
    if (no_stop == FALSE & "*" %in% seq){
      seq <- seq[1:min(which(seq == "*"))]
    }
    seq <- paste0(seq, collapse = sep)
    return(seq)
  }
}


R/utilities.R
#' Separate a character column into multiple columns with a regular expression
#'
#' @param .data A data frame.
#' @param col Column name to separate in.
#' @param into Names of new variables to create as character vector.
#' @param sep Separator between columns.
#' @return A data frame with the separated columns
#' @export
#'
#' @examples
#' df <- data.frame(a = c("first_second", "my_vec"))
#' df
#' separate_col(df, a, into = c("v1", "v2"))
separate_col <- function(.data, col, into, sep = "_"){
  var <- deparse(substitute(col))
  df <-
    .data %>%
    select(var) %>%
    pull() %>%
    strsplit(split = sep)
  df <-
    do.call(rbind,
            lapply(df, function(x)x)) %>%
    as.data.frame()
  .data[[var]] <- NULL
  names(df) <- into
  return(cbind(.data, df))
}


#' Find the antiparallel sequence
#'
#' @param x A character sequence of bases (A, T, C, G)
#'
#' @return A character sequence with the antiparalell sequence of `x`
#' @export
#'
#' @examples
#' seq <- "ATCG"
#' antiparallel(seq)
#'
#' seq2 <- c(seq, "AUGDF")
#' antiparallel(seq2)
antiparallel <- function(x){
  if(length(x) > 1){
    lapply(x, antiparallel)
  } else{
    x <- tidy_sequence(x)
    lapply(strsplit(x, "")[[1]], function(x){
      case_when(x == "A" ~ "T",
                x == "T" ~ "A",
                x == "C" ~ "G",
                x == "G" ~ "C",
                x == "U" ~ "A")
    }
    ) %>%
      as.character() %>%
      paste0(collapse = "")
  }
}

# Find the nth element/sequence
#' Extract the nth element of a numeric or character sequence
#'
#' @param x A numeric or character sequence
#' @param interval The nth  interval to select from the sequence
#'
#' @return A character or numeric vector depending on the `x` input
#' @export
#'
#' @examples
#' num <- 1:20
#' nth_element(num, interval = 5)
#'
#' let <- letters[1:20]
#' nth_element(let, interval = 2)
nth_element <- function(x, interval){
  x[seq(1, length(x), interval)]
}

#' Generate random codons with A, C, and G only
#'
#' @param n The number of random codons to generate.
#'
#' @return A character vector with the same length of `n`
#' @export
#'
#' @examples
#' rand_acg(n = 5)
rand_acg <- function(n){
  sapply(1:n, function(x){
    sample(c("A", "C", "G"), 3) %>% paste0(collapse = "")
  })
}

#' Split a character sequence systematically each `n` pieces
#'
#' @param x A character sequence/vector
#' @param n The number of pieces for each split
#'
#' @return A character vector
#' @export
#'
#' @examples
#' split_each("TESTSPLITEACH2", n = 2)
#'
#' # split a character vector each 3 letter
#' split_each(c("AAAAAABBBBBBB", "BAFASFDEASDFS"), n = 3)
#'
split_each <- function(x, n = 3){
  if(length(x) > 1){
    lapply(x, split_each, n)
  } else{
    regmatches(x, gregexpr(paste0(".{", n, "}"), x))[[1]]
  }
}

#' Tidy up a character vector
#'
#' The function removes all spaces, punctuation signals, and numbers from a
#' character vector, returning a 'letter only' upper-case sequence
#'
#' @param x A character vector.
#'
#' @return A character vector of the same length as `x`
#' @export
#'
#' @examples
#' messy_seq <- c("a 1, -C B", "AS.DF2D F33")
#' tidy_sequence(messy_seq)
tidy_sequence <- function(x){
  if(length(x) > 1){
    as.character(lapply(x, tidy_sequence))
  } else{
    str <- toupper(gsub("(?<=[a-z])(?=[A-Z])|[[:space:][:punct:]]+", "", x, perl = TRUE))
    return(gsub("[0-9]", "", str, perl = TRUE))
  }
}



#' @keywords internal
#' @importFrom poorman %>%
#' @export
#' @usage lhs \%>\% rhs
NULL


# check for bases
check_base <- function(x){
  if(length(x) > 1){
    lapply(x, check_base)
  } else{
    code <- str_split(x, "")[[1]]
    bases <- c("A", "U", "G", "C")
    if(all(code %in% bases) == FALSE){
      stop("'x' must contains only 'A', 'U', 'G', or 'C'.", call. = FALSE)
    }
  }
}

#' Check if a sequence is DNA or RNA
#'
#' `is_dna()` checks the vector `x` and returns `FALSE` if a Uracil (U) is found
#' in the sequence.
#' `is_rna()` checks the vector `x` and returns `FALSE` if a Thymine (T) is
#' found in the sequence.
#'
#' @param x A character vector
#' @param verbose Logical value indicating if a warning message should be
#'   returned.
#' @return A logical value indicating if a warning should be returned
#' @name util_is
#' @export
#'
#' @examples
#' is_dna("ATG")
#' is_rna(c("AUG", "TGA"))
is_dna <- function(x, verbose = TRUE){
  if (length(x) > 1) {
    as.logical(lapply(x, is_dna, verbose))
  } else{
    test <- str_detect(x, "U")
    if (test == TRUE) {
      if (isTRUE(verbose)) {
        warning("It seems that a RNA sequence was informed.", call. = FALSE)
      }
    }
    return(!test)
  }
}
#' @name util_is
#' @export
is_rna <- function(x, verbose = TRUE){
  if (length(x) > 1) {
    as.logical(lapply(x, is_rna, verbose))
  } else{
    test <- str_detect(x, "T")
    if (test == TRUE) {
      if (isTRUE(verbose)) {
        warning("It seems that a DNA sequence was informed.", call. = FALSE)
      }
    }
    return(!test)
  }
}



#' Highlight a pattern sequence in a vector
#'
#' The function highlights a pattern (e.g., "UAG") in a vector by putting a
#' replacement  (e.g., an underscore _) in all characters that don't match the
#' sequence
#'
#' @param x A character vector.
#' @param pattern A character vector indicating the pattern to highlight
#' @param replacement A replacement to replace all the characters that do not
#'   match the `pattern` with.
#'
#' @return A list with the location and highlighted pattern.
#' @export
#'
#' @examples
#' x <- rand_nuc_seq(n = 50, seed = 1)
#' x
#' highlight_pattern(x, pattern = "CCA")
#'
#' # or a vector
#' highlight_pattern(x, pattern = c("CCA", "GTA"))
#'
#' # vectorized
#'
#' x2 <-  rand_nuc_seq(n = 50, seed = 2)
#' x2
#' highlight_pattern(c(x, x2), pattern = "ATG")
highlight_pattern <- function(x,
                              pattern,
                              replacement = "_",
                              verbose = TRUE){
  if(length(x) > 1){
    a <- lapply(x, highlight_pattern, pattern, replacement, verbose)
  } else{
    loc <- str_locate_all(x, pattern)
    names(loc) <- pattern
    dots <- paste0(rep(replacement, nchar(x)), collapse = "")
    for(i in 1:length(loc)){
      start <- loc[[i]][,1]
      end <- loc[[i]][,2]
      nsub <- seq_along(start)
      for(j in nsub){
        str_sub(dots, start[j], end[j]) <- pattern[i]
      }
    }
    loc <- lapply(loc, function(x){
      as.data.frame(x)
    }) %>%
      bind_rows(.id = "pattern")
    if(isTRUE(verbose)){
      cat("============ Location ============\n")
      print(loc)
      cat("----------------------------------\n")
      message(dots)
    }
    invisible(list(location = loc, sequence = dots))
  }
}


R/rand_nuc_seq.R
#' Generate random nucleotide sequences
#'
#'Generate random sequences of nucleotides (i.e., A, T, C, and G) for DNA and A,
#'U, C, and G for RNA. (See details for more information).
#'
#'The random sequences are generated so that a start codon (AUG for RNA or ATG
#'for DNA) is inserted between the first and third bases in the chain. A stop
#'codon (UAA, UAG, or UGA) is randomly inserted at the end of the chain. A
#'message is returned with the position and stop codon used.
#'
#'
#' @param n The number of nucleotides.
#' @param type The type of sequence. It must be either `"dna"` to generate a DNA
#'   sequence, or `"rna"` to generate a RNA sequence
#' @param seed The seed for the random number generator . Set to ensure
#'   reproducibility
#'
#' @return A character vector.
#' @export
#'
#' @examples
#' # generate a DNA sequenc
#' rand_nuc_seq(n = 15, seed = 10)
#'
#' # generate a RNA sequenc
#' rand_nuc_seq(n = 15, type = "rna", seed = 10)
#'
rand_nuc_seq <- function(n = 30, type = "dna", seed = NULL){
  if(!type %in% c("dna", "rna")){
    warning("`type` must be one of 'dna' or 'rna'. Defaulting to 'dna'.", call. = FALSE)
    type <- "dna"
  }
  if(n < 10){
    stop("The number of bases must be greater than 10", call. = FALSE)
  }
  if(!is.null(seed)){
    set.seed(seed)
  }
  if(type == "dna"){
    start <- c("A", "T", "G")
  } else{
    start <- c("A", "U", "G")
  }
  if(type == "dna"){
    stops <- list(
      stop1 = c("T", "A", "A"),
      stop2 = c("T", "A", "G"),
      stop3 = c("T", "G", "A")
    )
  } else{
    stops <- list(
      stop1 = c("U", "A", "A"),
      stop2 = c("U", "A", "G"),
      stop3 = c("U", "G", "A")
    )
  }
  if(type == "dna"){
    samp <- sample(c("A", "T", "C", "G"), n, replace = TRUE)
  } else{
    samp <- sample(c("A", "U", "C", "G"), n, replace = TRUE)
  }
  # insert start codon in a random position
  # between first and fourth bases
  nin <- sample(1:3, 1)
  nfin <- nin + 2
  samp[nin:nfin] <- start
  # inserts a stop codon on a random position
  # From 4 bases after start codon up to two of the end
  free_seq <- nth_element((nfin + 4):(n-2), 3)
  if(length(free_seq) == 1){
    nin2 <- free_seq
  } else{
    nin2 <- sample(free_seq, 1)
  }
  trink <- (nin2-1 - nfin)%%3
  while(trink != 0 ){
    nin2 <- nin2 + 1
    trink <- (nin2 - nfin)%%3
  }
  nfin2 <- nin2 + 2
  # selects a stop codon at random
  nstop <- sample(1:3, 1)
  samp[nin2:nfin2] <- stops[[nstop]]

  samp <- paste0(samp, collapse = "")
  pre_met <- str_sub(samp, 1, nin - 1)
  met_stop <-
    str_sub(samp, nin, nin2 - 1) %>%
    split_each(n = 3)
    # replace possible stop codons within the chain with random codons
  met_stop <-
    met_stop %>%
    str_replace_all(stops$stop1 %>% paste0(collapse = ""), rand_acg(length(met_stop))) %>%
    str_replace_all(stops$stop2 %>% paste0(collapse = ""), rand_acg(length(met_stop))) %>%
    str_replace_all(stops$stop3 %>% paste0(collapse = ""), rand_acg(length(met_stop))) %>%
    paste0(collapse = "")
  pos_stop <- str_sub(samp, nin2, nchar(samp))
  samp <- str_c(pre_met, met_stop, pos_stop)
  message("Start codon  : ", paste0(nin, "-", nfin))
  message("Stop codon   : ", paste0(stops[[nstop]], collapse = ""))
  message("Position stop: ", paste0(nin2, "-", nfin2))
  return(samp)
}


R/gen_code.R
#' Tools for the genetic code
#'
#' * `aminoacid()` returns the aminoacid generated by the informed codon(s).
#' * `gen_code()` returns all the amino acids considering the translation table
#' used.
#'
#' @details The translation is performed according to the desired translation
#'   table. See more at
#'   \url{https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes}
#'
#'
#' 1. The Standard Code (default)
#' 2. The Vertebrate Mitochondrial Code
#' 3. The Yeast Mitochondrial Code
#' 4. The Mold, Protozoan, and Coelenterate Mitochondrial Code and the
#' Mycoplasma/Spiroplasma Code
#' 5. The Invertebrate Mitochondrial Code
#' 6. The Ciliate, Dasycladacean and Hexamita Nuclear Code
#' 9. The Echinoderm and Flatworm Mitochondrial Code
#' 10. The Euplotid Nuclear Code
#' 11. The Bacterial, Archaeal and Plant Plastid Code
#' 12. The Alternative Yeast Nuclear Code
#' 13. The Ascidian Mitochondrial Code
#' 14. The Alternative Flatworm Mitochondrial Code
#' 16. Chlorophycean Mitochondrial Code
#' 21. Trematode Mitochondrial Code
#' 22. Scenedesmus obliquus Mitochondrial Code
#' 23. Thraustochytrium Mitochondrial Code
#' 24. Rhabdopleuridae Mitochondrial Code
#' 25. Candidate Division SR1 and Gracilibacteria Code
#' 26. Pachysolen tannophilus Nuclear Code
#' 27. Karyorelict Nuclear Code
#' 28. Condylostoma Nuclear Code
#' 29. Mesodinium Nuclear Code
#' 30. Peritrich Nuclear Code
#' 31. Blastocrithidia Nuclear Code
#' 33. Cephalodiscidae Mitochondrial UAA-Tyr Code
#'
#'
#' @param codon A character vector (upper- or lower-case letters) with the 5'-3'
#'   codon (e.g., "AUG").
#' @param transl_table The translation table used. Defaults to 1 (standard
#'   genetic code). See more in `Details` section.
#' @param type The type of output. One of
#' * `"olc"` for one-letter code, e.g., "V".
#' * `"tlc"` for three-letter code, e.g., "Val".
#' * `"name"` for the complete name, e.g., "Valine".
#' @return
#' * `aminoacid()` returns a character vector in the format `code_letter_name` of the given aminoacid(s)
#' * `gen_code()` returns a data.frame object with the following columns.
#' * `fb` The first base of the codon.
#' * `sb` The second base of the codon.
#' * `tb` The third base of the codon.
#' * `codon` The codon (5'-3').
#' * `olc` The one letter code.
#' * `tlc` The three letter code.
#' * `name` The name of the aminoacid.
#'
#' The stop codons are depicted by `"*"`
#' @export
#' @name gen_code
#' @import poorman
#' @importFrom  stringr str_replace_all str_split str_locate str_locate_all str_sub str_detect str_c str_sub<-
#' @importFrom utils read.csv
#'
#' @examples
#' aac <- c("AUG", "CGA", "UGA")
#' aminoacid(aac)
#' aminoacid(aac, transl_table = 2)
#'
#' # complete genetic code
#' gen_code()
#'

aminoacid <- function(codon, transl_table = 1, type = "olc"){
  path <- system.file(package = "genomics")
  gcod <- read.csv(file =  paste0(path, "/gen_code.csv"), header = T, sep = ";")
  amino <- read.csv(file =  paste0(path, "/aminoacids.csv"), header = T, sep = ";")
  if(length(codon) > 1){
    as.character(do.call(rbind, lapply(codon, aminoacid, transl_table,  type)))
  } else{
    if(!type %in% c("olc", "tlc", "name")){
      stop("Type must be one 'olc', 'tlc', or 'name'.")
    }
    codon <- tidy_sequence(codon)
    check_base(codon)
    if(nchar(codon) != 3){
      stop("Codon must be a length of 3.", call. = FALSE)
    }
    olc <- gcod[which(codon == gcod$codon), which(colnames(gcod) == paste0("tt", transl_table))]
    amino[which(amino$olc == olc), type]
  }
}

#' @export
#' @name gen_code
gen_code <- function(transl_table = 1){
  bases <- c("A", "U", "G", "C")
  gcode <-
    expand.grid(fb = bases, sb = bases, tb = bases) %>%
    arrange(fb, sb, tb) %>%
    mutate(
      codon = paste0(fb, sb, tb),
      olc = aminoacid(codon, transl_table, type = "olc"),
      tlc = aminoacid(codon, transl_table, type = "tlc"),
      name = aminoacid(codon, transl_table, type = "name")
    )
  rownames(gcode) <- NULL
  return(gcode)
}
